### PR TITLE
Adapt deploy release prompts

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?

Follow-up: https://github.com/Shopify/cli/pull/3180

NOTES:
-  Add support for the first UX proposal. Final UX implementation in a following up [PR](https://github.com/Shopify/cli/pull/3205)
- `webhooks.api_version` not versioned yet but mandatory for backwards compatibility. The prompt always displays `webhooks` as a new field

### WHAT is this pull request doing?
- Refactor how the `deploy` commands generates the extensions diff content to display in the confirmation prompt:
  1. Apply the existing matching logic
  2. Build the breakdown info for the prompt
  3. Display the prompt
  4. In case the prompt is confirmed execute some post-confirmation actions (create new extensions)
- Add support for the new `Configuration` section displayed in the prompt
- Add the confirmation prompt UX to match the new [figma](https://www.figma.com/file/cTu9Ag52V4IdqZe1cij6kK/Versioned-app-config?type=design&node-id=9657-5889&mode=design&t=J4Va6ecVDXCIZhkk-0) 
- `release` command reuses most of the `deploy` confirmation prompt logic to generate its own prompt

### How to test your changes?
- Create a new `cli` app and run the `p shopify app config link --path /PATH/TO/YOUR/APP` command to link it to a new remote app. 
- Add/delete/modify some content from the `toml` file. Supported versioned configs:
  - app_home
  - app_proxy
  - point_of_sale
  - webhooks (declarative fields)
  - app_access
- Add/delete some regular extensions
- Run `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP` command
image

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/app/app.ts [hunk 1]
```diff
@@ -202,6 +210,12 @@ export function appIsLaunchable(app: AppInterface) {
   return Boolean(frontendConfig || backendConfig)
 }
 
+export function filterNonVersionedAppFields(app: AppInterface) {
+  return Object.keys(app.configuration).filter(
+    (fieldName) => !Object.keys(NonVersionedAppSchema.shape).concat('path').includes(fieldName),
+  )
+}
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/models/app/app.ts:217]: This should do it: `return Object.keys(VersionedAppSchema.parse(app.configuration))`

File: packages/app/src/cli/prompts/deploy-release.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+async function deployConfirmationPrompt({
+  appTitle,
+  extensionsContentPrompt: {extensionsInfoTable, deletedInfoTable},
+  configContentPrompt: {configInfoTable, deletedInfoTable: configDeletedInfoTable},
+  release,
+}: DeployConfirmationPromptOptions): Promise<boolean> {
+  const timeBeforeConfirmationMs = new Date().valueOf()
+  let confirmationResponse = true
+
+  let question = `Create a new version${appTitle ? ` of ${appTitle}` : ''}?`
```
Comment (isaacroldan) [inline packages/app/src/cli/prompts/deploy-release.ts:57]: let question = `${release ? 'Release' : 'Create'} a new version${appTitle ? ` of ${appTitle}` : ''}?`

or easier to read:
const type = release ? 'Release' : 'Create'
const title = appTitle ? ` of ${appTitle}` : ''
let question = `${type} a new version${title}?`

Or if you prefer the current way, i'm ok, just being nitpicky here

File: packages/app/src/cli/prompts/deploy-release.test.ts [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+import {SpyInstance, describe, expect, test, vi} from 'vitest'
+import * as ui from '@shopify/cli-kit/node/ui'
+
+describe('deployOrReleaseConfirmationPrompt', () => {
+  describe('when release', () => {
+    test('and force no prompt should be displayed and true returned', async () => {
+      // Given
+      const {extensionIdentifiersBreakdown, configExtensionIdentifiersBreakdown} = buildCompleteBreakdownInfo()
+      const renderConfirmationPromptSpyOn = vi.spyOn(ui, 'renderConfirmationPrompt')
+      const metadataSpyOn = vi.spyOn(metadata, 'addPublicMetadata').mockImplementation(async () => {})
```
Comment (isaacroldan) [inline packages/app/src/cli/prompts/deploy-release.test.ts:16]: why not completely mock `ui` and `metadata`?

File: packages/app/src/cli/services/context/breakdown-extensions.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  localRegistration: IdentifiersExtensions,
+  localSourceToCreate: LocalSource[],
+): ExtensionIdentifiersBreakdown {
+  const identifiersToUpdate = Object.keys(localRegistration)
+  const identifiersToCreate = localSourceToCreate.map((source) => source.localIdentifier)
+  return {
+    onlyRemote: [] as string[],
+    toCreate: [] as string[],
+    toUpdate: [...identifiersToUpdate, ...identifiersToCreate],
+    fromDashboard: [] as string[],
```
Comment (isaacroldan) [inline packages/app/src/cli/services/context/breakdown-extensions.ts:210]: Shouldn't `identifiersToCreate` be in the `toCreate` array?

File: packages/app/src/cli/services/release.ts [hunk 5]
```diff
@@ -30,10 +31,19 @@ interface ReleaseOptions {
...
-  if (!options.force) await confirmReleasePrompt(partnersApp.title, versionsDiff)
-
+  const {extensionIdentifiersBreakdown, versionDetails} = await extensionsIdentifiersReleaseBreakdown(
+    token,
+    partnersApp.apiKey,
+    options.version,
+    app.specifications ?? [],
+  )
+  const confirmed = await deployOrReleaseConfirmationPrompt({
+    extensionIdentifiersBreakdown,
```
Comment (isaacroldan) [inline packages/app/src/cli/services/release.ts:41]: why you don't use `configExtensionIdentifiersBreakdown` in this case? don't we want to prompt for that during release?

File: packages/app/src/cli/services/context/identifiers.test.ts [hunk 6]
```diff
@@ -1,196 +1,98 @@
...
-    // When
-    const got = ensureDeploymentIdsPresence(options([EXTENSION_A, EXTENSION_A_2], [FUNCTION_C]))
-
-    // Then
-    await expect(got).rejects.toThrow(/Deployment failed because this local project doesn't seem to match the app/)
-  })
-
-  test('throw an invalid environment error if there are pending remote matches', async () => {
-    // Given
-    vi.mocked(ensureExtensionsIds).mockResolvedValue(err('pending-remote'))
```
Comment (isaacroldan) [inline packages/app/src/cli/services/context/identifiers.test.ts:145]: what happened with all these possible errors?

File: packages/app/src/cli/models/app/app.ts [hunk 7]
```diff
@@ -98,56 +99,63 @@ const WebhooksSchemaWithDeclarative = WebhooksSchema.merge(DeclarativeWebhooksSc
...
-      })
-      .optional(),
-    extension_directories: zod.array(zod.string()).optional(),
-    web_directories: zod.array(zod.string()).optional(),
-  })
-  .strict()
+export const NonVersionedAppTopSchema = zod.object({
+  name: zod.string().max(30),
+  client_id: zod.string(),
+})
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/models/app/app.ts:105]: Just to avoid modifyig  the position of the fields inside the `toml` file. 

The final schema is created in this sentence 
```ts
export const AppSchema = NonVersionedAppTopSchema.merge(VersionedAppSchema).merge(NonVersionedAppBottomSchema).strict()
```
 That way when the config is written to the file it will be done in the same order.

Once the `name`, `access_scopes` and `auth` are versioned in a future PR we can simplify this to have just a `NonVersionedAppSchema`

File: packages/app/src/cli/models/app/app.ts [hunk 8]
```diff
@@ -98,56 +99,63 @@ const WebhooksSchemaWithDeclarative = WebhooksSchema.merge(DeclarativeWebhooksSc
...
+      automatically_update_urls_on_dev: zod.boolean().optional(),
+      dev_store_url: zod.string().optional(),
+    })
+    .optional(),
+  extension_directories: zod.array(zod.string()).optional(),
+  web_directories: zod.array(zod.string()).optional(),
+})
+export const NonVersionedAppSchema = NonVersionedAppTopSchema.merge(NonVersionedAppBottomSchema)
+
+export const VersionedAppSchema = zod.object({
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/models/app/app.ts:130]: This is completely achieved in the following [Generate app schema dynamically](https://github.com/Shopify/cli/pull/3191) PR

File: packages/app/src/cli/services/context/identifiers-extensions.test.ts [hunk 9]
```diff
@@ -520,311 +661,161 @@ describe('ensureExtensionsIds: excludes non uuid managed extensions as normal ex
...
+      extensionsNonUuidManaged: {},
     })
-
-    // Then
-    expect(deployConfirmationPrompt).not.toBeCalled()
-    expect(matchConfirmationPrompt).toBeCalled()
   })
 })
-
-describe('ensureExtensionsIds: Migrates extension', () => {
```
Comment (isaacroldan) [inline packages/app/src/cli/services/context/identifiers-extensions.test.ts:771]: why did you remove the migrate extensions tests?

Comment (isaacroldan) [reply packages/app/src/cli/services/context/identifiers-extensions.test.ts:771]: nevermind show them up top

File: packages/app/src/cli/models/app/app.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/models/app/app.ts:217]: I'm trying to get the fields in the current configuration file that are versioned.

`path`  is added to `app.configuration` content manually in other parts of the codebase (not included in the file) and we don't want to display it in the list of modified fields

> This should do it: return Object.keys(VersionedAppSchema.parse(app.configuration))

As mentioned in a previous comment, there's a following  [Generate app schema dynamically](https://github.com/Shopify/cli/pull/3191) PR so the fixed `VersionedAppSchema` is going to be removed because the schema will be generated dynamically from the `config` extension specs schemas

File: packages/app/src/cli/prompts/config.test.ts [hunk 10]
```diff
@@ -200,7 +200,10 @@ describe('confirmPushChanges', () => {
 
       configuration.name = 'app2'
       configuration.access_scopes = {scopes: 'read_themes, read_customers'}
-      configuration.webhooks.api_version = 'unstable'
+      configuration.webhooks = {
+        ...configuration.webhooks,
+        api_version: 'unstable',
+      }
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/prompts/config.test.ts:206]: `DEFAULT_CONFIG` includes `webhoooks` information so I have just modified the required field

File: packages/app/src/cli/prompts/deploy-release.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+    confirmationResponse = await renderConfirmationPrompt({
+      message: question,
+      infoTable,
+      confirmationMessage,
+      cancellationMessage: 'No, cancel',
+    })
+  }
+
+  const timeToConfirmOrCancelMs = new Date().valueOf() - timeBeforeConfirmationMs
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/prompts/deploy-release.ts:109]: This variable was existing previously so I didn't touch it. Anyway, I would just remove it and create an issue to dig into why it's not used instead of increase the size of this PR

File: packages/app/src/cli/prompts/deploy-release.ts (hunk 2 above)
Comment (alvaro-shopify) [reply packages/app/src/cli/prompts/deploy-release.ts:58]: I had applied this modification in the following [Add modifications to deploy release prompt](https://github.com/Shopify/cli/pull/3205) PR

File: packages/app/src/cli/services/context/breakdown-extensions.ts (hunk 4 above)
Comment (alvaro-shopify) [reply packages/app/src/cli/services/context/breakdown-extensions.ts:210]: As decided in the [figma](https://www.figma.com/file/cTu9Ag52V4IdqZe1cij6kK/Versioned-app-config?type=design&node-id=9657-5889&mode=design&t=J4Va6ecVDXCIZhkk-0) 

)

when you run the `deploy` command without release, the cli should display the list of local fields

File: packages/app/src/cli/services/context/identifiers.ts [hunk 12]
```diff
@@ -35,36 +34,36 @@ export interface LocalSource {
...
-    case 'pending-remote':
-    case 'invalid-environment':
-      throw new AbortError(
-        `Deployment failed because this local project doesn't seem to match the app "${appName}" in Shopify Partners.`,
-        `If you didn't intend to select this app, run ${
-          outputContent`${outputToken.packagejsonScript(packageManager, 'deploy', '--reset')}`.value
-        }
-• If this is the app you intended, check your local project and make sure
-  it contains the same number and types of extensions as the Shopify app
-  you've selected. You may need to generate missing extensions.`,
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/app/src/cli/services/context/identifiers.ts:65]: no, these error weren't used even before this PR

File: packages/app/src/cli/prompts/deploy-release.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+export async function deployOrReleaseConfirmationPrompt({
+  force,
+  extensionIdentifiersBreakdown,
+  configExtensionIdentifiersBreakdown,
+  appTitle,
+  release,
+}: DeployOrReleaseConfirmationPromptOptions) {
+  if (force) return true
+  const extensionsContentPrompt = await buildExtensionsContentPrompt(extensionIdentifiersBreakdown)
+  const configContentPrompt = await buildConfigContentPrompt(release, configExtensionIdentifiersBreakdown)
```
Comment (alexanderMontague) [inline packages/app/src/cli/prompts/deploy-release.ts:39]: This does not have to be an async function

File: packages/app/src/cli/prompts/deploy-release.ts [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  if (!configContentBreakdown)
+    return {
+      configInfoTable: {header: 'Configuration: ', items: []},
+      deletedInfoTable: undefined,
+    }
+
+  const {existingFieldNames, existingUpdatedFieldNames, newFieldNames, deletedFieldNames} = configContentBreakdown
+
+  const modifiedFieldNames = [
+    ...existingUpdatedFieldNames.map((field) => [field, {subdued: '(updated)'}]),
```
Comment (alexanderMontague) [inline packages/app/src/cli/prompts/deploy-release.ts:164]: There is a bug with `webhooks` always showing `(updated)` on the prompt table even when it has not been modified. This is due to `api_version` missing from the diff. This will now be included in the webhook module, so maybe we can include this now too

image


image

Comment (alexanderMontague) [reply packages/app/src/cli/prompts/deploy-release.ts:164]: semi related to this (I think), and this is more of a UX question potentially, but when you add declarative webhooks for the first time to your TOML, it says `webhooks (new)`. This is _technically_ not true as you will always have `api_version` under webhooks and that is already a mandatory field. What do you think of this? Should it just always say `updated`?

File: packages/app/src/cli/prompts/config.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/prompts/config.test.ts:206]: but `DEFAULT_CONFIG` only has `api_version` and since you are setting `api_version: 'unstable'` just below, having that line doesn't do anything no?

Or we can keep it for future usage if we are going to add more fields

File: packages/app/src/cli/prompts/deploy-release.ts (hunk 13 above)
Comment (alvaro-shopify) [reply packages/app/src/cli/prompts/deploy-release.ts:39]: yep, I will address this refactor in the final PR

File: packages/app/src/cli/prompts/deploy-release.ts (hunk 14 above)
Comment (alvaro-shopify) [reply packages/app/src/cli/prompts/deploy-release.ts:164]: I'm going to address the inclusion of the `api_version` as a versioned field in a new PR just to avoid doing a lot of `rebases`.

> semi related to this (I think), and this is more of a UX question potentially, but when you add declarative webhooks for the first time to your TOML, it says webhooks (new). This is technically not true as you will always have api_version under webhooks and that is already a mandatory field. What do you think of this? Should it just always say updated?

I will open a thread in the channel to discuss this topic. However from my point of view this will only happen the first time the user deploy the versioned config as after it the `breakdown info` will use the `registered config app modules`. The problem here is that adding the current api config to the logic to detect this scenario will make it really more complex to avoid this edge case
